### PR TITLE
Shelley forge should respect max block size override 

### PR DESCRIPTION
**Pull request by EncodePanda**
Addresses: https://github.com/input-output-hk/ouroboros-network/issues/3237 and https://github.com/input-output-hk/ouroboros-network/issues/3225

Also https://github.com/input-output-hk/ouroboros-network/issues/3239 was fixed in https://github.com/input-output-hk/ouroboros-network/pull/3238/commits/4b456e1a88aac78215e3aa4c11d8bd0754bc46b0

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/Block/Forging.hs [hunk 1]
```diff
@@ -159,11 +162,11 @@ data BlockForging m blk = BlockForging {
 -- It is possible to override this maximum transaction capacity with a lower
 -- value. We ignore higher values than the ledger state's max block size. Such
 -- blocks would be rejected by the ledger anyway.
-data MaxTxCapacityOverride
+data MaxTxCapacityOverride blk
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Block/Forging.hs:165]: The comment on this type and on the `MaxTxCapacityOverride` constructor now seem to focus on bytes too much; could use some rewording now that it's more general.

File: ouroboros-consensus/src/Ouroboros/Consensus/Block/Forging.hs [hunk 2]
```diff
@@ -173,16 +176,38 @@ data MaxTxCapacityOverride
...
-      MaxTxCapacityOverride txCap -> noOverride `min` txCap
+      MaxTxCapacityOverride txCap -> noOverride `minMeasure` txCap
   where
-    noOverride = maxTxCapacity ledger
+    noOverride = maxCapacity ledger
+    minMeasure x y
+      | lessEq @blk x y = x
+      | otherwise  = y
+
+
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Block/Forging.hs:195]: Extra newline

File: ouroboros-consensus/src/Ouroboros/Consensus/Block/Forging.hs [hunk 3]
```diff
@@ -173,16 +176,38 @@ data MaxTxCapacityOverride
...
+  -> Measure blk
 computeMaxTxCapacity ledger maxTxCapacityOverride = case maxTxCapacityOverride of
       NoMaxTxCapacityOverride     -> noOverride
-      MaxTxCapacityOverride txCap -> noOverride `min` txCap
+      MaxTxCapacityOverride txCap -> noOverride `minMeasure` txCap
   where
-    noOverride = maxTxCapacity ledger
+    noOverride = maxCapacity ledger
+    minMeasure x y
+      | lessEq @blk x y = x
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Block/Forging.hs:192]: This logic still seems too fragile to me. If the user accidentally sets one override above the ledger state's limit, then all of their overrides will be ignored.

**Review by nfrisby — COMMENTED** [review]
A few last bits to tie up.

File: ouroboros-consensus/src/Ouroboros/Consensus/Block/Forging.hs (hunk 3 above)
Comment (EncodePanda) [reply ouroboros-consensus/src/Ouroboros/Consensus/Block/Forging.hs:192]: Thinking to address it separately, issue defined here https://github.com/input-output-hk/ouroboros-network/issues/3239

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Mempool.hs [hunk 4]
```diff
@@ -42,12 +42,18 @@ import           Cardano.Binary (Annotator (..), FromCBOR (..),
...
 import           Ouroboros.Consensus.Block
 import           Ouroboros.Consensus.Ledger.Abstract
 import           Ouroboros.Consensus.Ledger.SupportsMempool
+import           Ouroboros.Consensus.Mempool.TxLimits
+import           Ouroboros.Consensus.Shelley.Eras (AllegraEra, AlonzoEra,
+                     MaryEra, ShelleyEra)
 import           Ouroboros.Consensus.Util (ShowProxy (..))
 import           Ouroboros.Consensus.Util.Condense
 
+
```
Comment (nfrisby) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Mempool.hs:55]: Extra newline

File: ouroboros-consensus/src/Ouroboros/Consensus/Block/Forging.hs (hunk 1 above)
Comment (nfrisby) [reply ouroboros-consensus/src/Ouroboros/Consensus/Block/Forging.hs:165]: We neglected to catch this one during our call

**Review by nfrisby — APPROVED** [review]
Approving; there are just a couple minor things left to do before merging:

  - update the comments around `MaxTxCapacityOverride`
  - remove a couple of extra newlines

File: ouroboros-consensus/src/Ouroboros/Consensus/Block/Forging.hs (hunk 1 above)
Comment (EncodePanda) [reply ouroboros-consensus/src/Ouroboros/Consensus/Block/Forging.hs:165]: done

**Comment by EncodePanda** [comment]
bors merge

**Pull request merged**